Terminal.Gui/View/ViewKeyboard.cs

```diff
@@ -1,464 +1,688 @@
+	/// <para>
+	/// Called after <see cref="OnKeyDown"/> and <see cref="OnKeyUp"/>.
+	/// </para>
+	/// <para>
+	/// SubViews can use the <see cref="KeyPressed"/> of their super view to intercept key presses.
+	/// </para>
+	/// <para>
+	/// Not all terminals support key distinct down/up notifications, Applications should avoid
+	/// depending on distinct KeyDown and KeyUp events and instead should use <see cref="KeyPressed"/>.
+	/// </para>
+	/// </remarks>
+	public virtual bool OnKeyPressed (KeyEventArgs keyEvent)
+	{
+		// fire event
+		KeyPressed?.Invoke (this, keyEvent);
+		return keyEvent.Handled;
+	}
 
-			var args = new KeyEventEventArgs (keyEvent);
-			KeyDown?.Invoke (this, args);
-			if (args.Handled) {
+	/// <summary>
+	/// Invoked when a key is pressed. Set <see cref="KeyEventArgs.Handled"/> to true to stop the key from
+	/// being processed by other views. Invoked after <see cref="KeyDown"/> and <see cref="KeyUp"/>.
+	/// </summary>
+	/// <remarks>
+	/// <para>
+	/// SubViews can use the <see cref="KeyPressed"/> of their super view to intercept key presses.
+	/// </para>
+	/// <para>
+	/// Not all terminals support key distinct down/up notifications, Applications should avoid
+	/// depending on distinct KeyDown and KeyUp events and instead should use <see cref="KeyPressed"/>.
+	/// </para>
+	/// <para>
+	/// See <see href="../docs/keyboard.md">for an overview of Terminal.Gui keyboard APIs.</see>
+	/// </para>
+	/// </remarks>
+	public event EventHandler<KeyEventArgs> KeyPressed;
+	#endregion Low-level Key handling
+
+	#region Key Bindings
+	/// <summary>
+	/// Gets the key bindings for this view.
+	/// </summary>
+	private Dictionary<Key, Command []> KeyBindings { get; set; } = new Dictionary<Key, Command []> ();
+	private Dictionary<Command, Func<bool?>> CommandImplementations { get; set; } = new Dictionary<Command, Func<bool?>> ();
+
+	/// <summary>
+	/// Low-level API called when a key is pressed to invoke any key bindings set on the view.
+	/// This is called after <see cref="OnKeyPressed(KeyEventArgs)"/>. 
+	/// </summary>
+	/// <remarks>
+	/// <para>
+	/// Fires the <see cref="InvokingKeyBindings"/> event.
+	/// </para>
+	/// <para>
+	/// See <see href="../docs/keyboard.md">for an overview of Terminal.Gui keyboard APIs.</see>
+	/// </para>
+	/// </remarks>
+	/// <param name="keyEvent">Contains the details about the key that produced the event.</param>
+	/// <returns><see langword="false"/> if the key press was not handled. <see langword="true"/> if
+	/// the keypress was handled and no other view should see it.</returns>
+	public virtual bool OnInvokeKeyBindings (KeyEventArgs keyEvent)
+	{
+		// fire event
+		InvokingKeyBindings?.Invoke (this, keyEvent);
+		if (keyEvent.Handled) {
+			return true;
+		}
+		var ret = InvokeKeyBindings (keyEvent);
+		if (ret != null && (bool)ret) {
+			return true;
+		}
+
+		foreach (var view in Subviews.Where (v => v is not Toplevel && v.Enabled && !v.HasFocus && v.KeyBindings.Count > 0)) {
```

Comment:
`View` base class doesn't have to knowing about a derived class `TopLevel`. The isolation principle is affected. If a derived class has a different behavior the it must overridden the related method and deal with the right actions. I know you want to get rid of the `TopLevel` but this derived view has specific functionalities that only have to belong to him.

Reply:
Another problem with this design is that are a lot of views that have `KeyBindings` that only must be run if they are focused. `TextView` is one of that and this call will invoke an action even if it isn't focused and thus having unexpected behavior. With this design force all the `TextView` methods called by `KeyBindings` to check if it's focused which is a nightmare.

Reply:
Before was more explicit what the `ProcessKey`, `ProcessHotKey` and `ProcessColdKey` were doing. Now leveraging the key bindings there is no way to distinguish between them. One solution is providing a key to the key bindings if it's processing a simple key press, a hotkey press or a cold key press. Would be good if @tznind to say what he thing about this because was he who create the key bindings.
Only to clarify a key press can be a single character or a combinations of keys that first is processed by the current focused view and if not handled the `TopLevel` will providing to a sub-view that need that keys for a hotkey or cold key processing. A hotkey press is a single character which is activated with the `Alt` key and could be the first character of a string or the character that is prefixed with a hotkey specifier, like `_`. If the current focused view doesn't handle it the `TopLevel` will provide it to a sub-view that needs them. A cold key press is a single character or a combinations of keys (shortcuts), like the `Enter` key that is used by the `Button` view if it's set as default, where if the current focused view doesn't handle it, the `TopLevel` will provide it to a sub-view that needs them and thus, the `Button` `Clicked` will be invoked.
It will be need three virtual methods anyway, `OnKeyPress`, `OnHotKeyPress` and `OnColdKeyPress`. Currentlly in the `v2_develop` branch the `Application.OnKeyPressed` is calling separately the `ProcessHotKey`, `ProcessKey` and the `ProcessColdKey`. In my opinion this only has to be needed to call once to `ProcessKey`. In the `View.ProcessKey` will provide in the `ProcessKey` for the current focused view and if not handled will provide in the `ProcessHotKey` for the current focused view and if not handled will provide in the `ProcessColdKey` for the current focused view and if not handled will provide in the `ProcessHotKey` for each sub-view the hotkey key and if not handled will provide in the `ProcessColdKey` for each sub-view the cold key.

Reply:
@tig I done the https://github.com/BDisp/Terminal.Gui/tree/v2_ensures-all-keys-for-most-focused-or-upwards branch only for prove what I said above. All the keys (`ProcessKey`, `ProcessHotKey` and `ProcessColdKey`) must be handled first by the most focused view, otherwise will navigate upwards until a superview wants it or return false. With this branch I found that a `Label` which can't focus, receives a `KeyPress` event and it's wrong. I had to set the `CanFocus` to true to allow that.